File: scripts/qualification/parent_modal_image.js
// Private diagnostic pixels from one original, non-input common dialog only.
// Never capture the desktop, dismiss a prompt, read input values or change prefs.
const done = arguments[arguments.length - 1];
const expected = arguments[0];
(async () => {
  const unavailable = {
    version: 2,
    kind: "unavailable",
    prompt: "unknown",
    message: "unknown",
    png: null,
  };
  if (typeof expected !== "string" || !/^[\x21-\x7e]{1,128}$/u.test(expected)) return unavailable;
  const w = window;
  const { NavigableManager } = ChromeUtils.importESModule(
    "chrome://remote/content/shared/NavigableManager.sys.mjs",
  );
  const original = w.gBrowser?.selectedBrowser;
  const validWindow = () =>
    original &&
    w.gBrowser.selectedBrowser === original &&
    original.currentURI?.spec === "about:blank" &&
    NavigableManager.getIdForBrowser(original) === expected &&
    w.document.documentElement.hasAttribute("window-modal-open");
  if (!validWindow()) return unavailable;
  const dialog = w.gDialogBox?.dialog;
  if (!dialog) return { ...unavailable, kind: "none" };
  const uri = "chrome://global/content/commonDialog.xhtml";
  const spotlight = "chrome://browser/content/spotlight.html";
  const initial = dialog._openedURL;
  if (initial != null && initial !== uri && initial !== spotlight)
    return { ...unavailable, kind: "other" };
  const frame = dialog._frame;
  const retained = () =>
    validWindow() &&
    w.gDialogBox.dialog === dialog &&
    dialog._frame === frame &&
    frame?.ownerDocument === w.document;
  if (!retained()) return unavailable;
  // SubDialog sets its URL only AFTER awaiting frame creation. Await the exact
  // original readiness promise, within the existing Marionette script deadline.
  const ready = dialog._dialogReady;
  if (!ready || typeof ready.then !== "function") return unavailable;
  await ready;
  if (!retained()) return unavailable;
  const opened = dialog._openedURL;
  if (initial != null && opened !== initial) return unavailable;
  if (opened !== uri && opened !== spotlight) return { ...unavailable, kind: "other" };
  const doc = frame.contentDocument;
  if (!doc || doc.documentURI !== opened) return unavailable;
  if (opened === spotlight) {
    // These are fixed built-in configuration identities, not dialog text, form
    // values, telemetry, arbitrary message IDs or an action/consent interface.
    const config = frame.contentWindow?.arguments?.[0];
    const id = config && Object.getOwnPropertyDescriptor(config, "id");
    const messages = new Map([
      ["NEW_USER_TOU_ONBOARDING", "new-user-terms"],
      ["PRE_ONBOARDING_SPLASH", "startup-splash"],
      ["AI_WINDOW_TOU_EXISTING_USERS_MODAL", "ai-window-terms"],
      ["LOGIN_STATUS_ADVISORY", "login-advisory"],
      ["BROWSER_BACKUP_OPTIN_SPOTLIGHT", "backup-optin"],
      ["FX_MR_106_UPGRADE", "upgrade"],
    ]);
    const message =
      id && Object.hasOwn(id, "value") ? (messages.get(id.value) ?? "unknown") : "unknown";
    if (
      !retained() ||
      frame.contentDocument !== doc ||
      doc.documentURI !== spotlight ||
      dialog._openedURL !== spotlight
    )
      return unavailable;
    return { ...unavailable, kind: "spotlight", message };
  }
  const type = doc.getElementById("commonDialog")?.getAttribute("windowtype");
  const types = new Map([
    ["prompt:alert", "alert"],
    ["prompt:alertCheck", "alertCheck"],
    ["prompt:confirm", "confirm"],
    ["prompt:confirmCheck", "confirmCheck"],
    ["prompt:confirmEx", "confirmEx"],
    ["prompt:prompt", "prompt"],
    ["prompt:promptUserAndPass", "promptUserAndPass"],
    ["prompt:promptPassword", "promptPassword"],
  ]);
  const prompt = types.get(type) ?? "unknown";
  const result = { ...unavailable, kind: "common", prompt };
  // Authentication/input/unknown dialogs get a fixed class only, never pixels.
  if (!["alert", "alertCheck", "confirm", "confirmCheck", "confirmEx"].includes(prompt)) {
    return result;
  }
  const validDialog = () =>
    validWindow() &&
    w.gDialogBox.dialog === dialog &&
    dialog._frame === frame &&
    frame.ownerDocument === w.document &&
    frame.contentDocument === doc &&
    doc.documentURI === uri &&
    doc.getElementById("commonDialog")?.getAttribute("windowtype") === type &&
    doc.getElementById("loginContainer")?.hidden === true &&
    doc.getElementById("password1Container")?.hidden === true;
  if (!validDialog()) return result;
  const r = frame.getBoundingClientRect();
  const values = [r.x, r.y, r.width, r.height, w.innerWidth, w.innerHeight];
  if (
    !values.every(Number.isFinite) ||
    r.x < 0 ||
    r.y < 0 ||
    r.width < 1 ||
    r.height < 1 ||
    r.width > 1600 ||
    r.height > 1200 ||
    r.x + r.width > w.innerWidth ||
    r.y + r.height > w.innerHeight
  )
    return result;
  let bitmap = null;
  let canvas = null;
  try {
    // DOM snapshot of the fixed iframe rectangle, not compositor/desktop readback.
    bitmap = await w.browsingContext.currentWindowGlobal.drawSnapshot(
      new DOMRect(r.x, r.y, r.width, r.height),
      1,
      "rgb(255,255,255)",
      { drawView: true },
    );
    const after = frame.getBoundingClientRect();
    if (!validDialog() || ["x", "y", "width", "height"].some((key) => after[key] !== r[key]))
      return result;
    canvas = w.document.createElementNS("http://www.w3.org/1999/xhtml", "canvas");
    canvas.width = Math.ceil(r.width);
    canvas.height = Math.ceil(r.height);
    canvas.getContext("2d").drawImage(bitmap, 0, 0);
    const data = canvas.toDataURL("image/png");
    const prefix = "data:image/png;base64,";
    if (data.startsWith(prefix) && data.length <= 1400000) result.png = data.slice(prefix.length);
    return result;
  } finally {
    try {
      if (bitmap) bitmap.close();
    } finally {
      if (canvas) {
        canvas.width = 0;
        canvas.height = 0;
      }
    }
  }
})().then(done, () =>
  done({ version: 2, kind: "unavailable", prompt: "unknown", message: "unknown", png: null }),
);

File: scripts/qualification/parent_modal_image.py
"""Private, bounded non-input dialog image; never an ordinary log or release artifact."""
import base64
import struct
import zlib
from .installed import ordinary

ERROR='owned modal image refused'
LIMIT=1024*1024
PROMPTS={'unknown','alert','alertCheck','confirm','confirmCheck','confirmEx','prompt','promptUserAndPass','promptPassword'}
NO_INPUT={'alert','alertCheck','confirm','confirmCheck','confirmEx'}
MESSAGES={'unknown','new-user-terms','startup-splash','ai-window-terms','login-advisory','backup-optin','upgrade'}


def png_bytes(encoded):
    if type(encoded) is not str or len(encoded)>1400000 or not encoded.isascii(): raise RuntimeError(ERROR)
    try: data=base64.b64decode(encoded,validate=True)
    except (ValueError,base64.binascii.Error): raise RuntimeError(ERROR) from None
    if (len(data)>LIMIT or data[:8]!=b'\x89PNG\r\n\x1a\n'
            or base64.b64encode(data).decode('ascii')!=encoded): raise RuntimeError(ERROR)
    offset=8;parts=[];kinds=[];size=None
    while offset<len(data) and len(kinds)<64:
        if len(data)-offset<12: raise RuntimeError(ERROR)
        length=struct.unpack('>I',data[offset:offset+4])[0]
        kind=data[offset+4:offset+8];end=offset+8+length
        if end+4>len(data) or kind not in {b'IHDR',b'IDAT',b'IEND',b'sRGB',b'gAMA',b'cHRM',b'pHYs',b'cICP'}: raise RuntimeError(ERROR)
        body=data[offset+8:end]
        if zlib.crc32(kind+body)!=struct.unpack('>I',data[end:end+4])[0]: raise RuntimeError(ERROR)
        if kind==b'IHDR':
            if kinds or length!=13: raise RuntimeError(ERROR)
            width,height,depth,color,compression,filtering,interlace=struct.unpack('>IIBBBBB',body)
            if not (1<=width<=1600 and 1<=height<=1200 and depth==8 and color in (2,6) and compression==filtering==interlace==0): raise RuntimeError(ERROR)
            size=(width*(3 if color==2 else 4)+1)*height
            stride=size//height
        elif not kinds: raise RuntimeError(ERROR)
        ancillary={b'sRGB':1,b'gAMA':4,b'cHRM':32,b'pHYs':9,b'cICP':4}
        if kind in ancillary and (kind in kinds or parts or length!=ancillary[kind]): raise RuntimeError(ERROR)
        if kind==b'IDAT': parts.append(body)
        if kind==b'IEND' and (length!=0 or end+4!=len(data)): raise RuntimeError(ERROR)
        kinds.append(kind);offset=end+4
    if offset!=len(data) or not kinds or kinds[-1]!=b'IEND' or not parts or size is None: raise RuntimeError(ERROR)
    try:
        decoder=zlib.decompressobj();raw=decoder.decompress(b''.join(parts),size+1)
    except zlib.error: raise RuntimeError(ERROR) from None
    if len(raw)!=size or not decoder.eof or decoder.unused_data or decoder.unconsumed_tail or any(raw[i]>4 for i in range(0,size,stride)): raise RuntimeError(ERROR)
    return data


def record(profile, response):
    if (type(response) is not dict or set(response)!={'version','kind','prompt','message','png'}
            or type(response['version']) is not int or response['version']!=2
            or type(response['kind']) is not str or response['kind'] not in {'unavailable','none','other','common','spotlight'}
            or type(response['prompt']) is not str or response['prompt'] not in PROMPTS
            or (response['kind']!='common' and response['prompt']!='unknown')
            or type(response['message']) is not str or response['message'] not in MESSAGES
            or (response['kind']!='spotlight' and response['message']!='unknown')): raise RuntimeError(ERROR)
    summary={'state':'observed','kind':response['kind'],'prompt':response['prompt'],'message':response['message'],'image_written':False}
    if response['png'] is None: return summary
    if response['kind']!='common' or response['prompt'] not in NO_INPUT: raise RuntimeError(ERROR)
    data=png_bytes(response['png'])
    ordinary(profile)
    if not profile.is_dir(): raise RuntimeError(ERROR)
    # The caller reserves its single observation before all image commands/I/O.
    # Fixed filename under the exclusively created test profile; no overwrite.
    with (profile/'owned-modal.private.png').open('xb') as out:
        if out.write(data)!=len(data): raise RuntimeError(ERROR)
        out.flush()
    summary['image_written']=True
    return summary
